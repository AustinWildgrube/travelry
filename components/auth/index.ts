export * from './RegistrationForm';
export * from './SetupConfetti';
export * from './SetupProfileDetails';
export * from './SetupProfileImage';