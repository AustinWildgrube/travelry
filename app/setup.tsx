import { Fragment, useState } from 'react';

import { Link } from 'expo-router';
import { SafeAreaView } from 'react-native-safe-area-context';

import { Confetti, SetupProfileDetails, SetupProfileImage } from '&/components/auth';
import { Button, Heading, Text } from '&/components/core';
import { CONFETTI_FALL_SPEED, CONFETTI_ITEM_HEIGHT, CONFETTI_ITEM_WIDTH } from '&/utils/constants';
import { makeStyles } from '&/utils/makeStyles';

export default function Setup(): JSX.Element {
  const styles = useStyles();
  const [step, setStep] = useState(0);

  return (
    <SafeAreaView style={styles.setup}>
      <Heading>Setup your profile</Heading>

      {step === 0 ? (
        <Fragment>
          <Text style={styles.setupSlogan}>Upload an image - it's worth a thousand words</Text>
          <SetupProfileImage setStep={setStep} />

          <Button onPress={() => setStep(1)} variant="link" style={styles.skipLink}>
            <Button.Text>Skip</Button.Text>
          </Button>

          <Confetti
            fallSpeed={CONFETTI_FALL_SPEED}
            flipSpeed={3}
            horizSpeed={50}
            itemColors={['#84a98c', '#ffb703', '#023047']}
            itemDimensions={{ height: CONFETTI_ITEM_HEIGHT, width: CONFETTI_ITEM_WIDTH }}
            itemTintStrength={0.8}
            numItems={100}
          />
        </Fragment>
      ) : (
        <Fragment>
          <Text style={styles.setupSlogan}>Tell the world a little about yourself</Text>
          <SetupProfileDetails />

          <Link href="/feed" asChild>
            <Button variant="link" style={styles.skipLink}>
              <Button.Text>Skip</Button.Text>
            </Button>
          </Link>
        </Fragment>
      )}
    </SafeAreaView>
  );
}

const useStyles = makeStyles(theme => ({
  setup: {
    flex: 1,
    padding: theme.space['4'],
  },
  setupSlogan: {
    marginTop: theme.space['1'],
    marginBottom: theme.space['6'],
  },
  skipLink: {
    marginTop: theme.space['3'],
    zIndex: -1,
  },
}));
