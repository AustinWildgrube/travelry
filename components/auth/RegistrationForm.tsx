import { useCallback, useMemo, useRef, useState, type MutableRefObject, type Ref } from 'react';
import { Keyboard, KeyboardAvoidingView, type TextInput, type TextInputProps } from 'react-native';

import BottomSheet, { BottomSheetBackdrop, BottomSheetScrollView } from '@gorhom/bottom-sheet';
import { type BottomSheetDefaultBackdropProps } from '@gorhom/bottom-sheet/lib/typescript/components/bottomSheetBackdrop/types';
import { zodResolver } from '@hookform/resolvers/zod';
import { Controller, useForm } from 'react-hook-form';
import DateTimePickerModal from 'react-native-modal-datetime-picker';
import { SafeAreaView } from 'react-native-safe-area-context';
import { z } from 'zod';

import { Button, Center, FormControl, Heading, Icon, Input, Text } from '&/components/core';
import { WarningIcon } from '&/components/core/Icons/Icons/Warning';
import { useAuth } from '&/contexts/Authentication';
import { makeStyles } from '&/utils/makeStyles';

type RegisterProps = {
  bottomSheetRef: MutableRefObject<BottomSheet | null>;
};

const RegistrationFormSchema = z.object({
  username: z
    .string()
    .refine(val => val.trim().length > 0, {
      message: 'A username is required',
    })
    .refine(val => !/\s/.test(val), {
      message: 'Your username cannot contain spaces',
    }),
  email: z
    .string()
    .email()
    .refine(val => val.trim().length > 0, {
      message: 'Your email is required',
    }),
  password: z.string().min(8, {
    message: 'Your password must be at least 8 characters long',
  }),
  birthdate: z.date().refine(val => {
    const today = new Date();
    const thirteenYearsAgo = new Date(today.getFullYear() - 13, today.getMonth(), today.getDate());
    return val <= thirteenYearsAgo;
  }, 'You must be 13 years or older to register'),
});

type RegistrationForm = z.infer<typeof RegistrationFormSchema>;

export function RegistrationForm({ bottomSheetRef }: RegisterProps): JSX.Element {
  const styles = useStyles();
  const { register } = useAuth();

  const snapPoints = useMemo(() => ['90%'], []);
  const birthdayInput = useRef<TextInput | null>(null);

  const [isDatePickerVisible, setIsDatePickerVisible] = useState(false);

  const {
    control,
    setValue,
    handleSubmit,
    formState: { errors },
  } = useForm<RegistrationForm>({
    resolver: zodResolver(RegistrationFormSchema),
  });

  const renderBackdrop = useCallback(
    (props: BottomSheetDefaultBackdropProps) => <BottomSheetBackdrop {...props} disappearsOnIndex={-1} appearsOnIndex={0} />,
    [],
  );

  const handleSheetChanges = (index: number): void => {
    if (index === -1 || 1) Keyboard.dismiss();
  };

  const showDatePicker = (): void => {
    setIsDatePickerVisible(true);
  };

  const hideDatePicker = (): void => {
    setIsDatePickerVisible(false);
  };

  const handleDateChange = (date: Date): void => {
    setValue('birthdate', date);
    birthdayInput.current?.setNativeProps({ value: date.toLocaleDateString() });
    hideDatePicker();
  };

  const submitRegistrationForm = (data: RegistrationForm): void => {
    const { email, password, username, birthdate } = data;
    register(email, password, username, birthdate);
  };

  return (
    <BottomSheet
      onChange={handleSheetChanges}
      backdropComponent={renderBackdrop}
      backgroundStyle={styles.bottomSheetBackground}
      handleIndicatorStyle={styles.bottomSheetHandle}
      snapPoints={snapPoints}
      ref={bottomSheetRef}
      index={-1}
      enablePanDownToClose>
      <SafeAreaView edges={['bottom', 'left', 'right']} style={styles.bottomSheetScroll}>
        <BottomSheetScrollView keyboardDismissMode="on-drag" showsVerticalScrollIndicator={true} style={styles.bottomSheetScroll}>
          <KeyboardAvoidingView behavior="position" keyboardVerticalOffset={20} style={styles.registrationForm}>
            <Heading size="2xl" style={styles.registrationHeader}>
              Itâ€™s Nice To Meet You
            </Heading>

            <Text style={styles.registrationSubHeader}>We're looking forward to seeing the memories you'll make and share!</Text>

            <Controller
              name="username"
              control={control}
              rules={{ required: true }}
              render={({ field: { onChange, onBlur, value } }) => (
                <FormControl isInvalid={!!errors.username} isRequired={true} style={styles.formControl}>
                  <FormControl.Label>
                    <FormControl.Label.Text>Username</FormControl.Label.Text>
                  </FormControl.Label>

                  <Input>
                    <Input.Input placeholder="Username" autoCapitalize="none" onChangeText={onChange} onBlur={onBlur} value={value} />
                  </Input>

                  <FormControl.Error>
                    <FormControl.Error.Icon>
                      <Icon as={WarningIcon} style={styles.formControlIcon} />
                    </FormControl.Error.Icon>

                    <FormControl.Error.Text>{errors.username?.message}</FormControl.Error.Text>
                  </FormControl.Error>
                </FormControl>
              )}
            />

            <Controller
              name="email"
              control={control}
              rules={{ required: true }}
              render={({ field: { onChange, onBlur, value } }) => (
                <FormControl isInvalid={!!errors.email} isRequired={true} style={styles.formControl}>
                  <FormControl.Label>
                    <FormControl.Label.Text>Email</FormControl.Label.Text>
                  </FormControl.Label>

                  <Input>
                    <Input.Input placeholder="Email" autoCapitalize="none" inputMode="email" onChangeText={onChange} onBlur={onBlur} value={value} />
                  </Input>

                  <FormControl.Error>
                    <FormControl.Error.Icon>
                      <Icon as={WarningIcon} style={styles.formControlIcon} />
                    </FormControl.Error.Icon>

                    <FormControl.Error.Text>{errors.email?.message}</FormControl.Error.Text>
                  </FormControl.Error>
                </FormControl>
              )}
            />

            <Controller
              name="password"
              control={control}
              rules={{ required: true }}
              render={({ field: { onChange, onBlur, value } }) => (
                <FormControl isInvalid={!!errors.password} isRequired={true} style={styles.formControl}>
                  <FormControl.Label>
                    <FormControl.Label.Text>Password</FormControl.Label.Text>
                  </FormControl.Label>

                  <Input>
                    <Input.Input placeholder="Password" autoCapitalize="none" type="password" onChangeText={onChange} onBlur={onBlur} value={value} />
                  </Input>

                  <FormControl.Error>
                    <FormControl.Error.Icon>
                      <Icon as={WarningIcon} style={styles.formControlIcon} />
                    </FormControl.Error.Icon>

                    <FormControl.Error.Text>{errors.password?.message}</FormControl.Error.Text>
                  </FormControl.Error>
                </FormControl>
              )}
            />

            <Controller
              name="birthdate"
              control={control}
              rules={{ required: true }}
              render={({ field: { onChange, onBlur, value } }) => (
                <FormControl isInvalid={!!errors.birthdate} isReadOnly={true} isRequired={true} style={styles.formControl}>
                  <FormControl.Label>
                    <FormControl.Label.Text>Birthday</FormControl.Label.Text>
                  </FormControl.Label>

                  <Input>
                    <Input.Input
                      placeholder="Birthday"
                      onPressOut={showDatePicker}
                      onChangeText={onChange}
                      onBlur={onBlur}
                      value={value ? value.toLocaleDateString('en-US') : ''}
                      ref={birthdayInput as Ref<TextInputProps> | undefined}
                    />
                  </Input>

                  <FormControl.Helper>
                    <FormControl.Helper.Text>Your birthday will not be shown publicly.</FormControl.Helper.Text>
                  </FormControl.Helper>

                  <FormControl.Error>
                    <FormControl.Error.Icon>
                      <Icon as={WarningIcon} style={styles.formControlIcon} />
                    </FormControl.Error.Icon>

                    <FormControl.Error.Text>{errors.birthdate?.message}</FormControl.Error.Text>
                  </FormControl.Error>
                </FormControl>
              )}
            />

            <Button onPress={handleSubmit(submitRegistrationForm)}>
              <Button.Text>Register</Button.Text>
            </Button>
          </KeyboardAvoidingView>
        </BottomSheetScrollView>

        <Center style={styles.legalWrapper}>
          <Text size="sm" style={styles.legalText}>
            By registering an account, you confirm that you agree to our Terms of Use and our Privacy Policy.
          </Text>
        </Center>

        <DateTimePickerModal
          mode="date"
          isVisible={isDatePickerVisible}
          onConfirm={handleDateChange}
          onCancel={hideDatePicker}
          testID="dateTimePicker"
        />
      </SafeAreaView>
    </BottomSheet>
  );
}

const useStyles = makeStyles(theme => ({
  bottomSheetBackground: {
    backgroundColor: theme.colors['white'],
  },
  bottomSheetHandle: {
    backgroundColor: theme.colors['dark50'],
  },
  bottomSheetScroll: {
    flex: 1,
  },
  registrationForm: {
    flex: 1,
    paddingHorizontal: theme.space['6'],
  },
  registrationHeader: {
    textAlign: 'center',
  },
  registrationSubHeader: {
    marginBottom: theme.space['8'],
    marginHorizontal: theme.space['4'],
    textAlign: 'center',
  },
  formControl: {
    marginBottom: theme.space['4'],
  },
  formControlIcon: {
    color: theme.colors['red500'],
    height: theme.space['3'],
    width: theme.space['3'],
  },
  legalWrapper: {
    marginHorizontal: theme.space['6'],
  },
  legalText: {
    textAlign: 'center',
  },
}));
