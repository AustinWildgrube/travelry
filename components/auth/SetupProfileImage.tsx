import { useState, type Dispatch, type SetStateAction } from 'react';

import { type CameraCapturedPicture } from 'expo-camera';
import { SaveFormat } from 'expo-image-manipulator';
import { launchImageLibraryAsync, MediaTypeOptions, type ImagePickerAsset } from 'expo-image-picker';
import { User } from 'lucide-react-native';

import { Camera } from '&/components/camera/Camera';
import { Avatar, Button, Icon, Spinner, VStack } from '&/components/core';
import { LoadingIndicator } from '&/components/shared/LoadingIndicator';
import { uploadUserAvatar } from '&/queries/user';
import { useUserStore } from '&/stores/user';
import { makeStyles } from '&/utils/makeStyles';
import { transformImageType } from '&/utils/transformImageType';

type SetupProfileProps = {
  setStep: Dispatch<SetStateAction<number>>;
};

export function SetupProfileImage({ setStep }: SetupProfileProps): JSX.Element {
  const styles = useStyles();

  const [showCamera, setShowCamera] = useState(false);
  const [imageUploading, setImageUploading] = useState(false);
  const [imageUri, setImageUri] = useState<string | null>(null);

  const user = useUserStore(state => state.user);
  if (!user) return <LoadingIndicator />;

  const pickAvatarImage = async (): Promise<void> => {
    let result = await launchImageLibraryAsync({
      mediaTypes: MediaTypeOptions.Images,
      allowsEditing: true,
      aspect: [1, 1],
      quality: 1,
      exif: true,
    });

    if (!result.canceled) {
      onSuccess(result.assets[0]);
    }
  };

  const onSuccess = (image: CameraCapturedPicture | ImagePickerAsset): void => {
    setImageUploading(true);

    transformImageType(image.uri, SaveFormat.JPEG).then(async transformedImage => {
      setImageUri(transformedImage.uri);
      await uploadUserAvatar(user.id, transformedImage);
      setStep(1);
    });
  };

  return (
    <VStack space="lg" style={styles.setupProfile}>
      {imageUri ? (
        <Avatar style={styles.avatar}>
          <Avatar.Image source={{ uri: imageUri }} />
        </Avatar>
      ) : (
        <Avatar style={styles.avatar}>
          <Icon as={User} style={styles.avatarIcon} />
        </Avatar>
      )}

      <Button onPress={pickAvatarImage} isDisabled={imageUploading} size="xl">
        {imageUploading && <Spinner style={styles.buttonSpinner} />}
        <Button.Text>Upload an Image</Button.Text>
      </Button>

      <Button onPress={() => setShowCamera(true)} isDisabled={imageUploading} size="xl" variant="outline">
        {imageUploading && <Spinner style={styles.buttonSpinner} />}
        <Button.Text>Take an Image</Button.Text>
      </Button>

      {showCamera && <Camera onSuccess={onSuccess} setIsShown={setShowCamera} />}
    </VStack>
  );
}

const useStyles = makeStyles(theme => ({
  setupProfile: {
    flex: 1,
    justifyContent: 'center',
  },
  avatar: {
    borderRadius: theme.radii['full'],
    height: theme.space['56'],
    width: theme.space['56'],
  },
  avatarIcon: {
    color: theme.colors['white'],
    height: theme.space['56'],
    width: theme.space['56'],
  },
  buttonSpinner: {
    color: theme.colors['white'],
    marginRight: theme.space['1'],
  },
}));
