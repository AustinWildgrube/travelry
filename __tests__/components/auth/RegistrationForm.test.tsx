import { fireEvent, screen, waitFor } from '@testing-library/react-native';

import { RegistrationForm } from '&/components/auth/RegistrationForm';

import { register, render } from '&/jest.setup';

describe('RegistrationForm', () => {
  beforeEach(() => {
    const bottomSheetRef = { current: null };
    render(<RegistrationForm bottomSheetRef={bottomSheetRef} />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    screen.toJSON();
    expect(screen).toMatchSnapshot();
  });

  it('should render all input fields', () => {
    const usernameInput = screen.getByRole('text', { name: /Username/i });
    const emailInput = screen.getByRole('text', { name: /Email/i });
    const passwordInput = screen.getByRole('text', { name: /Password/i });
    const birthdayInput = screen.getByRole('text', { name: /Birthday/i });

    expect(usernameInput).toBeDefined();
    expect(emailInput).toBeDefined();
    expect(passwordInput).toBeDefined();
    expect(birthdayInput).toBeDefined();
  });

  it('should hide the date picker when the cancel button is clicked', async () => {
    const birthdayInput = screen.getByPlaceholderText(/Birthday/i);
    fireEvent(birthdayInput, 'pressOut');

    const datePicker = await screen.findByTestId('dateTimePicker');
    expect(datePicker).toBeDefined();

    const cancelButton = await screen.findByRole('button', { name: /Cancel/i });
    fireEvent.press(cancelButton);

    await waitFor(() => {
      const closedDatePicker = screen.queryByTestId('dateTimePicker');
      expect(closedDatePicker).toBeNull();
    });
  });

  it('should set the birthday input value when a date is selected', async () => {
    const birthdayInput = screen.getByPlaceholderText(/Birthday/i);
    fireEvent(birthdayInput, 'pressOut');

    const datePicker = await screen.findByTestId('dateTimePicker');
    expect(datePicker).toBeDefined();

    const confirmButton = await screen.findByRole('button', { name: /Confirm/i });
    fireEvent.press(confirmButton);

    expect(birthdayInput.props.value).toBe(new Date().toLocaleDateString());
  });

  it('should error when the user is under the age of 13', async () => {
    const usernameInput = screen.getByPlaceholderText(/Username/i);
    const emailInput = screen.getByPlaceholderText(/Email/i);
    const passwordInput = screen.getByPlaceholderText(/Password/i);
    const birthdayInput = screen.getByPlaceholderText(/Birthday/i);

    fireEvent.changeText(usernameInput, 'testuser');
    fireEvent.changeText(emailInput, 'test@mail.com');
    fireEvent.changeText(passwordInput, 'password');
    fireEvent.changeText(birthdayInput, new Date(`${new Date().getFullYear() - 12}-01-01`));

    const registerButton = screen.getByRole('button', { name: /Register/i });
    fireEvent.press(registerButton);

    expect(await screen.findByText(/You must be 13 years or older to register/i)).toBeDefined();
    expect(register).not.toHaveBeenCalled();
  });

  it('should show an error messages if inputs are invalid', async () => {
    const submitButton = screen.getByRole('button', { name: /Register/i });
    fireEvent.press(submitButton);

    expect(await screen.findByText(/A username is required/i)).toBeDefined();
    expect(await screen.findByText(/Your email is required/i)).toBeDefined();
    expect(await screen.findByText(/A password is required/i)).toBeDefined();
    expect(await screen.findByText(/Your birthday is required/i)).toBeDefined();

    await waitFor(() => {
      expect(register).not.toHaveBeenCalled();
    });
  });

  it('should cal the register function when the form is submitted', async () => {
    const usernameInput = screen.getByPlaceholderText(/Username/i);
    const emailInput = screen.getByPlaceholderText(/Email/i);
    const passwordInput = screen.getByPlaceholderText(/Password/i);
    const birthdayInput = screen.getByPlaceholderText(/Birthday/i);

    fireEvent.changeText(usernameInput, 'testuser');
    fireEvent.changeText(emailInput, 'test@mail.com');
    fireEvent.changeText(passwordInput, 'password');
    fireEvent.changeText(birthdayInput, new Date('06/26/1996'));

    const submitButton = screen.getByRole('button');
    fireEvent.press(submitButton);

    await waitFor(() => expect(register).toHaveBeenCalled());
  });
});
