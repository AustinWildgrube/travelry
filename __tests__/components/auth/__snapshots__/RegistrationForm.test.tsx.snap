// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationForm should match the snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#ffffff",
        "flex": 1,
        "zIndex": 1,
      }
    }
  >
    <View
      backdropComponent={[Function]}
      backgroundStyle={
        {
          "backgroundColor": "#ffffff",
        }
      }
      enablePanDownToClose={true}
      handleIndicatorStyle={
        {
          "backgroundColor": "#242424",
        }
      }
      index={-1}
      onChange={[Function]}
      snapPoints={
        [
          "90%",
        ]
      }
    >
      <RCTScrollView
        keyboardDismissMode="on-drag"
        showsVerticalScrollIndicator={true}
        style={
          {
            "flex": 1,
          }
        }
      >
        <View>
          <View
            onLayout={[Function]}
            style={
              {
                "flex": 1,
                "padding": 12,
              }
            }
          >
            <View
              style={
                {
                  "bottom": 0,
                }
              }
            >
              <Text
                accessibilityRole="text"
                style={
                  {
                    "color": "#242424",
                    "fontFamily": "Inter_700Bold",
                    "fontSize": 32,
                    "fontWeight": "700",
                    "lineHeight": 40,
                    "marginTop": 8,
                    "textAlign": "center",
                  }
                }
              >
                Itâ€™s Nice To Meet You
              </Text>
              <Text
                accessibilityRole="text"
                style={
                  {
                    "color": "#242424",
                    "fontFamily": "Inter_400Regular",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "lineHeight": 20,
                    "marginBottom": 24,
                    "marginHorizontal": 24,
                    "marginTop": 8,
                    "textAlign": "center",
                  }
                }
              >
                We're looking forward to seeing the memories you'll make and share!
              </Text>
              <View
                style={
                  {
                    "marginBottom": 12,
                    "paddingHorizontal": 0,
                    "width": "100%",
                  }
                }
                testID="RNE__Input__view-wrapper"
              >
                <Text
                  accessibilityRole="text"
                  style={
                    {
                      "color": "#242424",
                      "fontFamily": "Inter_600SemiBold",
                      "fontSize": 14,
                      "fontWeight": "600",
                      "lineHeight": 20,
                      "marginLeft": 4,
                    }
                  }
                >
                  Username
                  <Text
                    accessibilityRole="text"
                    style={
                      {
                        "color": "#d46765",
                        "fontFamily": "Inter_400Regular",
                        "fontSize": 14,
                        "fontWeight": "400",
                        "lineHeight": 20,
                      }
                    }
                  >
                    *
                  </Text>
                </Text>
                <View
                  collapsable={false}
                  style={
                    {
                      "alignItems": "center",
                      "backgroundColor": "#fff",
                      "borderBottomWidth": 2,
                      "borderColor": "#d0d5dd",
                      "borderRadius": 8,
                      "borderStyle": "solid",
                      "borderWidth": 2,
                      "elevation": 2,
                      "flexDirection": "row",
                      "marginTop": 4,
                      "shadowColor": "rgba(16, 24, 40, .04)",
                      "shadowOffset": {
                        "height": 2,
                        "width": 0,
                      },
                      "shadowOpacity": 1,
                      "shadowRadius": 2,
                      "transform": [
                        {
                          "translateX": 0,
                        },
                      ],
                    }
                  }
                >
                  <TextInput
                    autoCapitalize="none"
                    editable={true}
                    onBlur={[Function]}
                    onChangeText={[Function]}
                    placeholder="Username"
                    placeholderTextColor="#86939e"
                    style={
                      {
                        "color": "#242424",
                        "flex": 1,
                        "fontSize": 16,
                        "fontWeight": "400",
                        "minHeight": 40,
                        "padding": 12,
                      }
                    }
                    testID="RNE__Input__text-input"
                    underlineColorAndroid="transparent"
                  />
                </View>
                <Text
                  style={
                    {
                      "color": "#8d9494",
                      "fontSize": 12,
                      "height": 0,
                      "margin": 0,
                      "marginBottom": 0,
                      "padding": 0,
                    }
                  }
                />
              </View>
              <View
                style={
                  {
                    "marginBottom": 12,
                    "paddingHorizontal": 0,
                    "width": "100%",
                  }
                }
                testID="RNE__Input__view-wrapper"
              >
                <Text
                  accessibilityRole="text"
                  style={
                    {
                      "color": "#242424",
                      "fontFamily": "Inter_600SemiBold",
                      "fontSize": 14,
                      "fontWeight": "600",
                      "lineHeight": 20,
                      "marginLeft": 4,
                    }
                  }
                >
                  Email
                  <Text
                    accessibilityRole="text"
                    style={
                      {
                        "color": "#d46765",
                        "fontFamily": "Inter_400Regular",
                        "fontSize": 14,
                        "fontWeight": "400",
                        "lineHeight": 20,
                      }
                    }
                  >
                    *
                  </Text>
                </Text>
                <View
                  collapsable={false}
                  style={
                    {
                      "alignItems": "center",
                      "backgroundColor": "#fff",
                      "borderBottomWidth": 2,
                      "borderColor": "#d0d5dd",
                      "borderRadius": 8,
                      "borderStyle": "solid",
                      "borderWidth": 2,
                      "elevation": 2,
                      "flexDirection": "row",
                      "marginTop": 4,
                      "shadowColor": "rgba(16, 24, 40, .04)",
                      "shadowOffset": {
                        "height": 2,
                        "width": 0,
                      },
                      "shadowOpacity": 1,
                      "shadowRadius": 2,
                      "transform": [
                        {
                          "translateX": 0,
                        },
                      ],
                    }
                  }
                >
                  <TextInput
                    autoCapitalize="none"
                    editable={true}
                    inputMode="email"
                    onBlur={[Function]}
                    onChangeText={[Function]}
                    placeholder="Email"
                    placeholderTextColor="#86939e"
                    style={
                      {
                        "color": "#242424",
                        "flex": 1,
                        "fontSize": 16,
                        "fontWeight": "400",
                        "minHeight": 40,
                        "padding": 12,
                      }
                    }
                    testID="RNE__Input__text-input"
                    underlineColorAndroid="transparent"
                  />
                </View>
                <Text
                  style={
                    {
                      "color": "#8d9494",
                      "fontSize": 12,
                      "height": 0,
                      "margin": 0,
                      "marginBottom": 0,
                      "padding": 0,
                    }
                  }
                />
              </View>
              <View
                style={
                  {
                    "marginBottom": 12,
                    "paddingHorizontal": 0,
                    "width": "100%",
                  }
                }
                testID="RNE__Input__view-wrapper"
              >
                <Text
                  accessibilityRole="text"
                  style={
                    {
                      "color": "#242424",
                      "fontFamily": "Inter_600SemiBold",
                      "fontSize": 14,
                      "fontWeight": "600",
                      "lineHeight": 20,
                      "marginLeft": 4,
                    }
                  }
                >
                  Password
                  <Text
                    accessibilityRole="text"
                    style={
                      {
                        "color": "#d46765",
                        "fontFamily": "Inter_400Regular",
                        "fontSize": 14,
                        "fontWeight": "400",
                        "lineHeight": 20,
                      }
                    }
                  >
                    *
                  </Text>
                </Text>
                <View
                  collapsable={false}
                  style={
                    {
                      "alignItems": "center",
                      "backgroundColor": "#fff",
                      "borderBottomWidth": 2,
                      "borderColor": "#d0d5dd",
                      "borderRadius": 8,
                      "borderStyle": "solid",
                      "borderWidth": 2,
                      "elevation": 2,
                      "flexDirection": "row",
                      "marginTop": 4,
                      "shadowColor": "rgba(16, 24, 40, .04)",
                      "shadowOffset": {
                        "height": 2,
                        "width": 0,
                      },
                      "shadowOpacity": 1,
                      "shadowRadius": 2,
                      "transform": [
                        {
                          "translateX": 0,
                        },
                      ],
                    }
                  }
                >
                  <TextInput
                    autoCapitalize="none"
                    editable={true}
                    onBlur={[Function]}
                    onChangeText={[Function]}
                    placeholder="Password"
                    placeholderTextColor="#86939e"
                    secureTextEntry={true}
                    style={
                      {
                        "color": "#242424",
                        "flex": 1,
                        "fontSize": 16,
                        "fontWeight": "400",
                        "minHeight": 40,
                        "padding": 12,
                      }
                    }
                    testID="RNE__Input__text-input"
                    underlineColorAndroid="transparent"
                  />
                </View>
                <Text
                  style={
                    {
                      "color": "#8d9494",
                      "fontSize": 12,
                      "height": 0,
                      "margin": 0,
                      "marginBottom": 0,
                      "padding": 0,
                    }
                  }
                />
              </View>
              <View
                style={
                  {
                    "marginBottom": 12,
                    "paddingHorizontal": 0,
                    "width": "100%",
                  }
                }
                testID="RNE__Input__view-wrapper"
              >
                <Text
                  accessibilityRole="text"
                  style={
                    {
                      "color": "#242424",
                      "fontFamily": "Inter_600SemiBold",
                      "fontSize": 14,
                      "fontWeight": "600",
                      "lineHeight": 20,
                      "marginLeft": 4,
                    }
                  }
                >
                  Birthday
                  <Text
                    accessibilityRole="text"
                    style={
                      {
                        "color": "#d46765",
                        "fontFamily": "Inter_400Regular",
                        "fontSize": 14,
                        "fontWeight": "400",
                        "lineHeight": 20,
                      }
                    }
                  >
                    *
                  </Text>
                </Text>
                <View
                  collapsable={false}
                  style={
                    {
                      "alignItems": "center",
                      "backgroundColor": "#fff",
                      "borderBottomWidth": 2,
                      "borderColor": "#d0d5dd",
                      "borderRadius": 8,
                      "borderStyle": "solid",
                      "borderWidth": 2,
                      "elevation": 2,
                      "flexDirection": "row",
                      "marginTop": 4,
                      "shadowColor": "rgba(16, 24, 40, .04)",
                      "shadowOffset": {
                        "height": 2,
                        "width": 0,
                      },
                      "shadowOpacity": 1,
                      "shadowRadius": 2,
                      "transform": [
                        {
                          "translateX": 0,
                        },
                      ],
                    }
                  }
                >
                  <TextInput
                    onBlur={[Function]}
                    onChangeText={[Function]}
                    onPressOut={[Function]}
                    placeholder="Birthday"
                    placeholderTextColor="#86939e"
                    style={
                      {
                        "color": "#242424",
                        "flex": 1,
                        "fontSize": 16,
                        "fontWeight": "400",
                        "minHeight": 40,
                        "padding": 12,
                      }
                    }
                    testID="RNE__Input__text-input"
                    underlineColorAndroid="transparent"
                    value=""
                  />
                </View>
                <Text
                  style={
                    {
                      "color": "#8d9494",
                      "fontSize": 12,
                      "margin": 5,
                      "marginBottom": 0,
                    }
                  }
                >
                  Your birthday will not be shown publicly.
                </Text>
              </View>
              <View
                style={
                  [
                    {
                      "overflow": "hidden",
                    },
                    {
                      "borderRadius": 8,
                    },
                    undefined,
                    false,
                  ]
                }
                testID="RNE_BUTTON_WRAPPER"
              >
                <View
                  accessibilityRole="button"
                  accessibilityState={
                    {
                      "busy": false,
                      "checked": undefined,
                      "disabled": false,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "opacity": 1,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "alignItems": "center",
                        "backgroundColor": "#84a98c",
                        "borderBottomWidth": 4,
                        "borderColor": "#6A8770",
                        "borderRadius": 8,
                        "borderStyle": "solid",
                        "borderWidth": 1,
                        "flexDirection": "row",
                        "justifyContent": "center",
                        "padding": 12,
                        "paddingHorizontal": 14,
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "color": "#fff",
                          "fontSize": 18,
                          "fontWeight": "600",
                          "paddingVertical": 1,
                          "textAlign": "center",
                        }
                      }
                    >
                      Register
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
      <View
        style={
          {
            "alignItems": "center",
            "paddingTop": 12,
          }
        }
      >
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#242424",
              "fontFamily": "Inter_400Regular",
              "fontSize": 12,
              "fontWeight": "400",
              "lineHeight": 20,
              "marginBottom": 24,
              "textAlign": "center",
              "width": "75%",
            }
          }
        >
          By registering an account, you confirm that you agree to our Terms of Use and our Privacy Policy.
        </Text>
      </View>
      <Modal
        animationType="none"
        hardwareAccelerated={false}
        isVisible={false}
        onHide={[Function]}
        transparent={true}
        visible={false}
      />
    </View>
  </View>
</View>
`;
